### PR TITLE
awsbatch: download AmazonLinux image from ECR rather than Docker Hub

### PR DESCRIPTION
**Pull request by demartinofra**
### Description of changes

* The change from `amazonlinux:latest` to `amazonlinux:2` is done to make sure we alway pull the latest for AL 2
* `docker: 18` removed from install/runtime-versions in CodeBuild buildspec because it is already available in the used container image and not needed anymore
* AL docker image is now pulled from ECR when the architecture is x86. ECR repos do not contain yet multi-arch images that support ARM. 
* Each partition is mapped to a single ECR repo hosted in a specific region. Currently there is no API to retrieve the registry id for AL images in the various regions, hence we keep this static mapping in the template for now.
* A new policy has been added to the CodeBuild role to allow access to the public ECR repo. This needs to be added to our docs.
* A fallback mechanism is put in place so that if the pull from ECR fails there is always a fallback to Docker Hub
* Integration tests have been extended to verify that the image is correctly pulled from ECR whenever possible

### Testing

I ran the following tests:
```
{%- import 'common.jinja2' as common -%}
test-suites:
  schedulers:
    test_awsbatch.py::test_awsbatch:
      dimensions:
        - regions: ["us-east-2"]
          instances: ["m6g.xlarge"]
          oss: ["alinux2"]
          schedulers: ["awsbatch"]
        - regions: {{ common.REGIONS_COMMERCIAL }}
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["awsbatch"]
        - regions: {{ common.REGIONS_CHINA }}
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["awsbatch"]
        - regions: {{ common.REGIONS_GOVCLOUD }}
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["awsbatch"]
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,13 @@ CHANGELOG
 - Remove CloudFormation DescribeStacks API call from AWS Batch Docker entrypoint. This removes the possibility of job
   failures due to CloudFormation throttling.
 
+**CHANGES**
+
+- Pull Amazon Linux Docker images from ECR when building docker image for `awsbatch` scheduler. This only applies to
+  images built for `x86` architecture.
```
Comment (lukeseawalker) [inline CHANGELOG.md:15]: worth adding about the fallback mechanism?

**Review by lukeseawalker — APPROVED** [review]

File: cli/src/pcluster/resources/batch/docker/pull-alinux-image.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+set -eux
```
Comment (enrico-usai) [inline cli/src/pcluster/resources/batch/docker/pull-alinux-image.sh:2]: let's add pipefail too: `set -euxo pipefail`

File: cli/src/pcluster/resources/batch/docker/upload-docker-images.sh [hunk 3]
```diff
@@ -1,17 +1,20 @@
 #!/usr/bin/env bash
-set -eu
+set -eux
```
Comment (enrico-usai) [inline cli/src/pcluster/resources/batch/docker/upload-docker-images.sh:2]: same here.

File: cli/src/pcluster/resources/batch/docker/pull-alinux-image.sh (hunk 2 above)
Comment (demartinofra) [reply cli/src/pcluster/resources/batch/docker/pull-alinux-image.sh]: done

File: cli/src/pcluster/resources/batch/docker/upload-docker-images.sh (hunk 3 above)
Comment (demartinofra) [reply cli/src/pcluster/resources/batch/docker/upload-docker-images.sh]: done

**Review by enrico-usai — APPROVED** [review]

**Pull request merged**